Add alarm switches to sonos component

AaronDavidSchneider opened the pull request:
## Breaking Change:
adding automatically discovered switches for sonos alarms to sonos integration

<!-- What is breaking and why we have to break it. Remove this section only if it was NOT a breaking change. -->

## Description:
many people control their sonos alarms using home assistant. However up until now, it was not possible to retreave the the actual alarm state. The sonos component only exposed a service to update the alarm states.

This pull request adds automatically discovered switches for every sonos alarm.

It will only affect those users who actually use sonos alarms.

**Related issue (if applicable):** 
Discussions about sonos alarms:
https://community.home-assistant.io/t/alarm-clock-with-sonos-philips-hue/23421
https://community.home-assistant.io/t/how-to-read-sonos-alarm-state-and-time/98998
https://community.home-assistant.io/t/control-sonos-alarm-with-philips-hue-dimmer/98730

Documentation PR: home-assistant/home-assistant.io#11277

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

balloob commented:
It seems like you have some commits in this branch that are not yours, please clean it up.

AaronDavidSchneider committed:
cleaned up pull request

AaronDavidSchneider commented:
> It seems like you have some commits in this branch that are not yours, please clean it up.

Sorry for that... should be cleaned up now

AaronDavidSchneider committed:
tried to fix CL issues

pvizeli commented on homeassistant/components/sonos/switch.py:
```diff
@@ -0,0 +1,216 @@
+        }
+        super().__init__()
+
+    async def async_update(self, now=None):
```
Soco is not async. Please use only `update`

AaronDavidSchneider replied:
changed it

AaronDavidSchneider committed:
removed async in update

AaronDavidSchneider committed:
cosmetics:
- change name of switches
- add entity ids
- changed volume from percent to fraction

AaronDavidSchneider reviewed (commented):
cosmetics: renamings, attributes changed.

amelchio commented:
The discovery loop should not be duplicated between media_player and switch. It should probably be moved to `sonos/__init__.py` and forward its discoveries to each platform.

Also, the Sonos integration is currently "local push". This switch addition is "local poll". It would be nice to maintain the "push" feature. That would probably mean a move/generalization of the event subscription as well.

AaronDavidSchneider commented:
Thankyou for your insight @amelchio.
I thought about moving the discovery to __init.py__ too. However this would significantly enlarge the pull request.
I can try this tomorrow.

amelchio commented:
Yeah, that is a solid task by itself. It would probably make sense to do that move as a preparation in a separate PR.

AaronDavidSchneider commented:
So you say merge this one first, tidy up later @amelchio?

amelchio commented:
By "preparation", I meant to do the refactor first. Promises of tidying up later don't always come true (that's a general comment, not implying anything about you in particular :-).

AaronDavidSchneider commented:
@amelchio:
I looked in the source code again to find a different solution.
Putting the discovered players somewhere into `hass.data[SONOS_DOMAIN]` is the only way I can see this working.

That would mean in practice:
discovery of players is done in `__init__.py`, discovered players are passed to `hass.data[SONOS_DOMAIN]`, in the async_setup_entry those are read and initialized.

What do you think about it?

Regarding the event subsciption:
I don't fully understand that part of the code. I don't think that I would be able to rewrite the eventsubscription such that it automatically pushes updates when a sonos alaram is changed externally (I don't even know how homeassistant would know, if not asking once every few seconds).

amelchio commented:
The discovery loop is a bit more complicated than that. It marks players available/unavailable during runtime when they start/stop responding to the discovery broadcasts.

We currently use `homeassistant.helpers.dispatcher` to get service calls from the component into the media_player platform.

I think something similar could be used to signal available/unavailable from the component to both media_player and switch platforms. This would ensure that alarm switches go unavailable when their owning player does.

Adding event subscription support for alarms is not straightforward, it would involve some low-level network sniffing to learn the data structures used.

(Sorry for the slow response. Also, I didn't look into the details of alarms so I may be a bit off)

AaronDavidSchneider commented:
Thankyou for your response and feedback @amelchio!
As this is way above my head (and I wouldnt have the time to dig into this issue deeper), its propably best to close this PR. I will simply use the code of my PR as custom_component.
